src/lib.rs

```diff
@@ -0,0 +1,270 @@
+use log::trace;
+use std::{collections, fs, path};
+
+pub struct OverdropConf {
+    dirs: Vec<path::PathBuf>,
+}
+
+impl OverdropConf {
+    pub fn new(
+        base_dirs: &Vec<String>,
+        config_path: &str,
+    ) -> Self {
+        let mut dirs = Vec::with_capacity(base_dirs.len());
+        for bdir in base_dirs {
+            let mut dpath = path::PathBuf::from(bdir);
+            dpath.push(config_path.clone());
+            dirs.push(dpath);
+        }
+        Self { dirs }
+    }
+
+    pub fn new_full(
+        root_dir: &str,
+        base_dirs: &Vec<String>,
+        config_path: &str,
+        version: Option<String>,
+    ) -> Self {
+        let mut dirs = Vec::with_capacity(base_dirs.len());
+        let version = version.unwrap_or("".to_string());
+        for bdir in base_dirs {
+            let mut dpath = path::PathBuf::from(root_dir);
+            dpath.push(bdir);
+            dpath.push(format!("{}{}", config_path, version));
+            dirs.push(dpath);
+        }
+        Self { dirs }
+    }
+
+    pub fn scan_unique_files(
+        &self,
+        allow_hidden: bool,
+        allowed_extensions: Option<&Vec<&str>>,
+    ) -> failure::Fallible<collections::BTreeMap<String, path::PathBuf>> {
```

Comment:
Unless I missed something, I think current code never returns an Error.

Reply:
:+1: - I have now changed this to return the `collections::BTreeMap<String, path::PathBuf>` - this should always be returned.